### PR TITLE
Improve out of order op processing

### PR DESCRIPTION
**Pull request by GaryWilber**
Consider the following scenario:

1. Client loads a summary @ seq 150 and connects to the document
2. Server sends the client their join op @ seq 160
    - This causes the client to add that op into `pending` and call `fetchMissingDeltas` to ask for ops 151 - 159
    - Lets say the storage api is not returning any deltas. At this point the client keeps retrying that call until it gets them
3. Server sends the client ops 151 - 159 through the WebSocket
4. Client processes ops 151 - 159

At this point the client has all the ops it needs to be fully connected. However `deltaManager` did not process op 160, which remains in `pending`. So the client is stuck waiting for `fetchMissingDeltas` to complete. This PR makes the client process that pending op so it won't get stuck.

Changes:
- Process pending ops after successfully processing an inbound op
- Stop trying to fetch deltas from storage if the delta manager already caught up
- Replace `catchUpCore` method with `processPendingOps` method that only processes pending ops

### CODE REVIEW

File: packages/loader/container-loader/src/deltaManager.ts [hunk 1]
```diff
@@ -835,6 +835,14 @@ export class DeltaManager
                 }
             }
 
+            if (to !== undefined && this.lastQueuedSequenceNumber >= to) {
+                // the client caught up while we were trying to fetch ops from storage
+                // bail out since we no longer need to request these ops
+                callback([]);
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:841]: I do not think making an empty call is required, or we do it in other places.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 2]
```diff
@@ -835,6 +835,14 @@ export class DeltaManager
...
+                telemetryEvent.end({ deltasRetrievedTotal, requests });
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:842]: having some additional property here to differentiate this path vs. any other path might be useful later in debugging.

File: packages/loader/container-loader/src/deltaManager.ts (hunk 1 above)
Comment (GaryWilber) [reply packages/loader/container-loader/src/deltaManager.ts:841]: The callback invokes `this.refreshDelayInfo(this.deltaStorageDelayId);`. If we skip this callback, that won't be called. Is that okay?

Comment (GaryWilber) [reply packages/loader/container-loader/src/deltaManager.ts]: I updated it to not pass anything so only `refreshDelayInfo` is invoked.

Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts]: It's important. I think it's cleaner for fetchMissingDeltas() to call it after it's done.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 3]
```diff
@@ -731,7 +731,7 @@ export class DeltaManager
         telemetryEventSuffix: string,
         fromInitial: number,
         to: number | undefined,
-        callback: (messages: ISequencedDocumentMessage[]) => void) {
+        callback: (messages?: ISequencedDocumentMessage[]) => void) {
```
Comment (vladsud) [inline packages/loader/container-loader/src/deltaManager.ts:734]: Even if you keep calling callback() on line 841, I'd rather pass [] and keep messages as required argument - feels a bit cleaner

Comment (GaryWilber) [reply packages/loader/container-loader/src/deltaManager.ts]: I switched it back and made `fetchMissingDeltas` call `refreshDelayInfo` like you mentioned.

**Review by vladsud — APPROVED** [review]

File: packages/loader/container-loader/src/deltaManager.ts (hunk 1 above)
Comment (GaryWilber) [reply packages/loader/container-loader/src/deltaManager.ts]: I assume it still has to call `refreshDelayInfo` in every callback too?

Comment (vladsud) [reply packages/loader/container-loader/src/deltaManager.ts]: Yes

**Pull request merged**